File: implementations/python/src/reifinator/generator.py
"""Core generation engine."""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Any

from reifinator.content import BaseContentGenerator, BuiltinInterpolator
from reifinator.context import CONTEXT_SCRIPT_NAME, load_context_script
from reifinator.output import OutputDirectory
from reifinator.resolution import (
    BracketResolver,
    GenerationTracker,
    PlaceholderResolver,
)

logger = logging.getLogger(__name__)

# Items to always skip during directory walking
SKIP_NAMES = {"__pycache__"}


class Generator:
    """Template-structure-driven code and directory generator.

    Walks a template directory tree, resolves placeholders, invokes content
    generators, and produces a mirrored output tree.
    """

    def __init__(
        self,
        template_dir: str | Path,
        output_dir: str | Path | None = None,
        output_stage_dir: str | Path | None = None,
        output_dest_dir: str | Path | None = None,
        content_generators: list[BaseContentGenerator] | None = None,
        placeholder_resolver: PlaceholderResolver | None = None,
        context_script_name: str = CONTEXT_SCRIPT_NAME,
        debug: bool = False,
    ):
        self.template_dir = Path(template_dir)
        self.context_script_name = context_script_name
        self.debug = debug
        self.placeholder_resolver = placeholder_resolver or BracketResolver()

        # Build extension -> generator mapping
        # Always include the built-in interpolator
        self._content_generators: dict[str, BaseContentGenerator] = {}
        builtin = BuiltinInterpolator()
        self._content_generators[builtin.extension] = builtin
        for gen in content_generators or []:
            self._content_generators[gen.extension] = gen

        # Resolve output directories
        self._resolve_output_dirs(output_dir, output_stage_dir, output_dest_dir)

    def _resolve_output_dirs(
        self,
        output_dir: str | Path | None,
        output_stage_dir: str | Path | None,
        output_dest_dir: str | Path | None,
    ) -> None:
        has_single = output_dir is not None
        has_stage = output_stage_dir is not None
        has_dest = output_dest_dir is not None

        if has_single and (has_stage or has_dest):
            raise ValueError(
                "Cannot specify both output_dir and output_stage_dir/output_dest_dir. "
                "Use output_dir for 1-stage or output_stage_dir + output_dest_dir for 2-stage."
            )
        if has_stage != has_dest:
            raise ValueError(
                "output_stage_dir and output_dest_dir must both be specified for 2-stage output."
            )
        if not has_single and not has_stage:
            raise ValueError(
                "No output directory configured. "
                "Specify output_dir (1-stage) or output_stage_dir + output_dest_dir (2-stage)."
            )

        if has_single:
            self.output_write_dir = Path(output_dir)
            self.output_dest_dir: Path | None = None
        else:
            self.output_write_dir = Path(output_stage_dir)
            self.output_dest_dir = Path(output_dest_dir)

    def run(self, context: dict[str, Any] | None = None) -> None:
        """Run the generation.

        Args:
            context: Initial root context. Merged with any context scripts found.
        """
        if not self.template_dir.is_dir():
            raise FileNotFoundError(f"Template directory does not exist: {self.template_dir}")

        root_context = context or {}
        tracker = GenerationTracker()
        output_root = OutputDirectory(self.output_write_dir)
        output_root.path.mkdir(parents=True, exist_ok=True)

        self._process_directory(self.template_dir, output_root, root_context, tracker)
        tracker.raise_if_unresolved()

    def _process_directory(
        self,
        input_path: Path,
        output_dir: OutputDirectory,
        parent_ctx: dict[str, Any],
        tracker: GenerationTracker,
    ) -> None:
        # Load context script if present
        script_path = input_path / self.context_script_name
        contexts = load_context_script(script_path, parent_ctx)

        for idx, ctx in enumerate(contexts):
            merged_ctx = {**parent_ctx, **ctx}

            if self.debug:
                self._write_context_log(output_dir.path, idx, merged_ctx)

            for item in input_path.iterdir():
                # Skip context scripts, cache dirs
                if item.name == self.context_script_name or item.name in SKIP_NAMES:
                    continue

                resolution = self.placeholder_resolver.resolve(item.name, merged_ctx)

                # Track items with expressions
                if resolution.has_expressions:
                    tracker.register_expression_item(item, resolution.expressions)
                    if resolution.success:
                        tracker.mark_resolved(item)

                # Skip if resolution failed
                if not resolution.success:
                    logger.debug("Skipping %s: placeholder not in current context", item.name)
                    continue

                output_name = resolution.resolved

                if item.is_dir():
                    new_output_dir = output_dir.create_dir(output_name)
                    self._process_directory(item, new_output_dir, merged_ctx, tracker)

                elif self._is_template_file(item):
                    # Duplicate prevention: skip non-substituted files in later iterations
                    if not resolution.was_substituted and idx > 0:
                        continue

                    ext = self._get_template_extension(item)
                    gen = self._content_generators[ext]
                    content = gen.generate(item, merged_ctx)
                    final_name = output_name.removesuffix(ext)
                    output_dir.write_file(final_name, content)

                else:
                    # Static file — copy as-is
                    # Duplicate prevention for static files in iterated directories
                    if not resolution.was_substituted and idx > 0:
                        continue
                    output_dir.copy_file(output_name, item)

    def _is_template_file(self, path: Path) -> bool:
        return any(path.name.endswith(ext) for ext in self._content_generators)

    def _get_template_extension(self, path: Path) -> str:
        for ext in self._content_generators:
            if path.name.endswith(ext):
                return ext
        return ""

    def _write_context_log(
        self, output_path: Path, idx: int, context: dict[str, Any]
    ) -> None:
        import pprint

        log_content = f"# Merged context for iteration {idx}\n{pprint.pformat(context)}"
        log_path = output_path / f".gen_context_{idx}.log"
        log_path.write_text(log_content)

File: implementations/typescript/src/generator.ts
/**
 * Core generation engine.
 */

import { existsSync, mkdirSync, readdirSync, statSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import { BaseContentGenerator, BuiltinInterpolator } from "./content.js";
import { CONTEXT_SCRIPT_NAMES, OTHER_CONTEXT_SCRIPTS, loadContextScript } from "./context.js";
import type { ContextDict } from "./context.js";
import { OutputDirectory } from "./output.js";
import {
  BracketResolver,
  GenerationTracker,
  type PlaceholderResolver,
} from "./resolution.js";

/** Items to always skip during directory walking. */
const SKIP_NAMES = new Set([
  "__pycache__",
  "node_modules",
  ".DS_Store",
]);

export interface GeneratorOptions {
  templateDir: string;
  outputDir?: string;
  outputStageDir?: string;
  outputDestDir?: string;
  contentGenerators?: BaseContentGenerator[];
  placeholderResolver?: PlaceholderResolver;
  contextScriptNames?: string[];
  debug?: boolean;
}

export class Generator {
  private readonly templateDir: string;
  private readonly outputWriteDir: string;
  private readonly outputDestDir: string | null;
  private readonly contentGenerators: Map<string, BaseContentGenerator>;
  private readonly placeholderResolver: PlaceholderResolver;
  private readonly contextScriptNames: string[];
  private readonly debug: boolean;

  constructor(options: GeneratorOptions) {
    this.templateDir = options.templateDir;
    this.placeholderResolver =
      options.placeholderResolver ?? new BracketResolver();
    this.contextScriptNames =
      options.contextScriptNames ?? CONTEXT_SCRIPT_NAMES;
    this.debug = options.debug ?? false;

    // Build extension -> generator mapping (built-in always included)
    this.contentGenerators = new Map<string, BaseContentGenerator>();
    const builtin = new BuiltinInterpolator();
    this.contentGenerators.set(builtin.extension, builtin);
    for (const gen of options.contentGenerators ?? []) {
      this.contentGenerators.set(gen.extension, gen);
    }

    // Resolve output directories
    const [writeDir, destDir] = this.resolveOutputDirs(
      options.outputDir,
      options.outputStageDir,
      options.outputDestDir,
    );
    this.outputWriteDir = writeDir;
    this.outputDestDir = destDir;
  }

  private resolveOutputDirs(
    outputDir?: string,
    outputStageDir?: string,
    outputDestDir?: string,
  ): [string, string | null] {
    const hasSingle = outputDir !== undefined;
    const hasStage = outputStageDir !== undefined;
    const hasDest = outputDestDir !== undefined;

    if (hasSingle && (hasStage || hasDest)) {
      throw new Error(
        "Cannot specify both outputDir and outputStageDir/outputDestDir. " +
          "Use outputDir for 1-stage or outputStageDir + outputDestDir for 2-stage.",
      );
    }
    if (hasStage !== hasDest) {
      throw new Error(
        "outputStageDir and outputDestDir must both be specified for 2-stage output.",
      );
    }
    if (!hasSingle && !hasStage) {
      throw new Error(
        "No output directory configured. " +
          "Specify outputDir (1-stage) or outputStageDir + outputDestDir (2-stage).",
      );
    }

    if (hasSingle) {
      return [outputDir!, null];
    }
    return [outputStageDir!, outputDestDir!];
  }

  async run(context?: ContextDict): Promise<void> {
    if (!existsSync(this.templateDir) || !statSync(this.templateDir).isDirectory()) {
      throw new Error(
        `Template directory does not exist: ${this.templateDir}`,
      );
    }

    const rootContext = context ?? {};
    const tracker = new GenerationTracker();
    const outputRoot = new OutputDirectory(this.outputWriteDir);
    mkdirSync(this.outputWriteDir, { recursive: true });

    await this.processDirectory(
      this.templateDir,
      outputRoot,
      rootContext,
      tracker,
    );
    tracker.raiseIfUnresolved();
  }

  private async processDirectory(
    inputPath: string,
    outputDir: OutputDirectory,
    parentCtx: ContextDict,
    tracker: GenerationTracker,
  ): Promise<void> {
    const contexts = await loadContextScript(
      inputPath,
      this.contextScriptNames,
      parentCtx,
    );

    for (let idx = 0; idx < contexts.length; idx++) {
      const ctx = contexts[idx];
      const mergedCtx = { ...parentCtx, ...ctx };

      if (this.debug) {
        this.writeContextLog(outputDir.path, idx, mergedCtx);
      }

      const items = readdirSync(inputPath);
      for (const itemName of items) {
        // Skip context scripts (own and other implementations), cache dirs
        if (
          this.contextScriptNames.includes(itemName) ||
          OTHER_CONTEXT_SCRIPTS.includes(itemName) ||
          SKIP_NAMES.has(itemName)
        ) {
          continue;
        }

        const itemPath = join(inputPath, itemName);
        const resolution = this.placeholderResolver.resolve(
          itemName,
          mergedCtx,
        );

        // Track items with expressions
        if (resolution.expressions.length > 0) {
          tracker.registerExpressionItem(itemPath, resolution.expressions);
          if (resolution.success) {
            tracker.markResolved(itemPath);
          }
        }

        // Skip if resolution failed
        if (!resolution.success) {
          continue;
        }

        const outputName = resolution.resolved!;
        const stat = statSync(itemPath);

        if (stat.isDirectory()) {
          const newOutputDir = outputDir.createDir(outputName);
          await this.processDirectory(
            itemPath,
            newOutputDir,
            mergedCtx,
            tracker,
          );
        } else if (this.isTemplateFile(itemName)) {
          // Duplicate prevention: skip non-substituted files in later iterations
          if (!resolution.wasSubstituted && idx > 0) continue;

          const ext = this.getTemplateExtension(itemName)!;
          const gen = this.contentGenerators.get(ext)!;
          const content = gen.generate(itemPath, mergedCtx);
          const finalName = outputName.slice(
            0,
            outputName.length - ext.length,
          );
          outputDir.writeFile(finalName, content);
        } else {
          // Static file — copy as-is
          // Duplicate prevention for static files in iterated directories
          if (!resolution.wasSubstituted && idx > 0) continue;
          outputDir.copyFile(outputName, itemPath);
        }
      }
    }
  }

  private isTemplateFile(name: string): boolean {
    for (const ext of this.contentGenerators.keys()) {
      if (name.endsWith(ext)) return true;
    }
    return false;
  }

  private getTemplateExtension(name: string): string | null {
    for (const ext of this.contentGenerators.keys()) {
      if (name.endsWith(ext)) return ext;
    }
    return null;
  }

  private writeContextLog(
    outputPath: string,
    idx: number,
    context: ContextDict,
  ): void {
    const logContent = `# Merged context for iteration ${idx}\n${JSON.stringify(context, null, 2)}`;
    const logPath = join(outputPath, `.gen_context_${idx}.log`);
    writeFileSync(logPath, logContent, "utf-8");
  }
}
